tests/rptest/utils/si_utils.py

```diff
@@ -333,59 +341,178 @@ def is_close_size(actual_size, expected_size):
+        else:
+            return any(key.endswith(t) for t in self.topic_manifest_paths)
 
     def is_partition_manifest(self, o: ObjectMetadata) -> bool:
-        return o.key.endswith('/manifest.json') and any(
-            tn in o.key for tn in self.topic_names)
+        return o.key.endswith(
+            '/manifest.json') and self._match_partition_manifest(o.key)
 
     def is_topic_manifest(self, o: ObjectMetadata) -> bool:
-        return any(o.key.endswith(t) for t in self.topic_manifest_paths)
+        return self._match_topic_manifest(o.key)
 
     def is_segment(self, o: ObjectMetadata) -> bool:
         try:
-            return parse_s3_segment_path(o.key).ntp.topic in self.topic_names
+            parsed = parse_s3_segment_path(o.key)
         except Exception:
             return False
+        else:
+            if self.topic_names is not None:
+                return parsed.ntp.topic in self.topic_names
+            else:
+                return True
 
     def path_matches_any_topic(self, path: str) -> bool:
-        return any(t in path for t in self.topic_names)
+        return self._match_partition_manifest(path)
 
 
-class S3Snapshot:
-    def __init__(self, expected_topics: Sequence[TopicSpec], client: S3Client,
-                 bucket: str, logger):
-        self.logger = logger
-        self.bucket = bucket
-        self.client = client
-        self.expected_topics = expected_topics
-        self.path_matcher = PathMatcher(self.expected_topics)
-        self.objects = self.client.list_objects(self.bucket)
+class BucketViewState:
+    """
+    Results of a full listing of the bucket, if listed is True, or
+    partial results if listed is False
+    """
+    def __init__(self):
+        self.listed = False
         self.segment_objects = 0
         self.ignored_objects = 0
-
         self.partition_manifests = {}
+
+
+class BucketView:
+    """
+    A caching view onto an object storage bucket, that knows how to construct paths
+    for manifests, and can also scan the bucket for aggregate segment info.
+
+    For directly fetching a manifest or manifest-derived info for a partition,
+    this will avoid doing object listings.  Access to properties that require
+    object listings (like the object counts) will list the bucket and then cache
+    the result.
+    """
+    def __init__(self, redpanda, topics: Optional[Sequence[TopicSpec]] = None):
+        """
+
+        :param redpanda: a RedpandaService, whose SISettings we will use
+        :param topics: optional list of topics to filter result to
+        """
+        self.redpanda = redpanda
+        self.logger = redpanda.logger
+        self.bucket = redpanda.si_settings.cloud_storage_bucket
+        self.client = redpanda.cloud_storage_client
+        self.path_matcher = PathMatcher(topics)
+
+        # Cache built on demand from admin API
+        self._ntp_to_revision = None
+
+        self._state = BucketViewState()
+
+    def refresh(self):
```

Comment:
nit: maybe `clear()` or `reset()` or `reset_cache()`? As is, it's a little ambiguous whether this automatically refetches all elements in the cache

Reply:
Fair point, changed this to reset.